test/server/server_test.cc

```diff
@@ -174,17 +174,23 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, ServerInstanceImplTest,
 
 TEST_P(ServerInstanceImplTest, LifecycleNotifications) {
   initialize("test/server/node_bootstrap.yaml");
-  bool got_startup_notification = false, got_shutdown_notification = false;
-  server_->registerCallback(ServerLifecycleNotifier::Stage::Startup,
-                            [&got_startup_notification, this] {
-                              got_startup_notification = true;
-                              server_->shutdown();
-                            });
+  bool startup = false, shutdown = false, shutdown_with_completion = false;
+  server_->registerCallback(ServerLifecycleNotifier::Stage::Startup, [&startup, this] {
+    startup = true;
+    server_->shutdown();
+  });
+  server_->registerCallback(ServerLifecycleNotifier::Stage::ShutdownExit,
+                            [&shutdown] { shutdown = true; });
+  Event::Dispatcher& dispatcher = server_->dispatcher();
   server_->registerCallback(ServerLifecycleNotifier::Stage::ShutdownExit,
-                            [&got_shutdown_notification] { got_shutdown_notification = true; });
+                            [&shutdown_with_completion, &dispatcher](Event::PostCb completion_cb) {
+                              shutdown_with_completion = true;
+                              dispatcher.post(completion_cb);
+                            });
   server_->run();
-  EXPECT_TRUE(got_startup_notification);
-  EXPECT_TRUE(got_shutdown_notification);
+  EXPECT_TRUE(startup);
+  EXPECT_TRUE(shutdown);
+  EXPECT_TRUE(shutdown_with_completion);
```

Comment:
I don't see this as testing that the variant that takes a completion_cb also delays server shutdown.  Would you add a test for that?  Also for startup, if it's intended to delay startup.

Reply:
It does test the blocking behavior. If the callback which sets shutdown_with_completion does not post completion_cb to the dispatcher the test will fail with a timeout.

Reply:
Given the test as written, I don't think there's any different in behavior between an implementation that blocks until the callback has been posted and one that doesn't.   Is there something I'm misunderstanding?

Reply:
Ok, I've improved the unit test - PTAL.